### ISSUE TITLE
Constraint Name Length

### ISSUE THREAD

**Issue by RFSH** [informatics-isi-edu/ermrest#135]
Recently one of the emrestjs test cases start failing and it is showing a 409 error. After investigating the problem, it seems like the constraint name for key has a maximum length. 

After changing the constraint name to something shorter, the request passes successfully and doesn't throw the 409 error.

**Comment by chiragsanghvi** [comment]
Here's the CURL request for the same. 

```sh
curl -X POST \
-d '{"comment":null,"kind":"table","entityCount":0,"keys":[{"comment":null,"annotations":{},"unique_columns":["column_without_display_annotation"]}],"foreign_keys":[],"table_name":"table_with_titlecase_without_underlinespace","schema_name":"schema_with_underlinespace_without_titlecase","column_definitions":[{"comment":null,"name":"column_without_display_annotation","default":null,"nullok":false,"type":{"typename":"serial4"},"annotations":{}},{"comment":null,"name":"column_with_underlinespace_without_titlecase","default":null,"nullok":false,"type":{"typename":"serial4"},"annotations":{"tag:misd.isi.edu,2015:display":{"name_style":{"title_case":false,"underline_space":true}}}}],"annotations":{"tag:misd.isi.edu,2015:display":{"name_style":{"title_case":true,"underline_space":false}}}}' \
-H "Cookie: webauthn=23PHiyiZx0WIewUdiUeyjNvx" \
https://dev.isrd.isi.edu:443/ermrest/catalog/18093/schema/schema_with_underlinespace_without_titlecase/table
```

**RESPONSE**
```sh
409 Conflict
The request conflicts with the state of the server. cannot drop index schema_with_underlinespace_without_titlecase.table_with_titlecase_without_underlinespace_column_without_disp because constraint table_with_titlecase_without_underlinespace_column_without_disp on table schema_with_underlinespace_without_titlecase.table_with_titlecase_without_underlinespace requires it
HINT:  You can drop constraint table_with_titlecase_without_underlinespace_column_without_disp on table schema_with_underlinespace_without_titlecase.table_with_titlecase_without_underlinespace instead.
```

**Comment by karlcz** [comment]
OK, it sounds like the name is being truncated and so what we think will be distinct names are instead colliding.

This will take some ugly revisions to how ERMrest names indices and constraints if we want to fully insulate ourselves from this.  I'm not sure it is worth it, as we still expose many other Postgres identifier length limitations as well.

Since you have a workaround (use shorter names) I am going to leave this as a low priority bug...

**Comment by karlcz** [comment]
FYI I've developed some fixes for this but as part of an ermrest feature branch that will not be merged for a while yet.  It seemed like it would interact with too many lines of code that have changed for this branch, so I didn't want to do it in master and then sort out the ugly merge conflicts...

The new code will reject user-supplied identifiers that exceed 63 bytes after UTF-8 encoding, and our internally-generated identifiers for DB indices will also adapt to make sure we don't exceed 63 bytes when concatenating several names together.  It will fall back to using truncated hashes of names if necessary to save space.

**Comment by karlcz** [comment]
The fixes described above are part of PR #152

**Comment by karlcz** [comment]
I don't think there is anybody around who really plans to review this, and the ID length enforcement is now tested in the new test/resttest/degenerate.py LongIdentifiers test class, so I am just going to close this old issue...
